### PR TITLE
Add SoftmaxCrossEntropyLossInternal to Support Dynamic ignore_index Input

### PR DESCRIPTION
**Pull request by iK1D**
Add SoftmaxCrossEntropyLossInternal and its gradient to support dynamic ignore_index input.

### CODE REVIEW

**Review by mrry — APPROVED** [review]
This is great, thanks Vincent!

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 1]
```diff
@@ -627,6 +627,44 @@ def run_step(model, input):
         _test_helpers.assert_values_are_close(ort_prediction, pt_prediction)
         _test_helpers.assert_gradients_match_and_reset_gradient(ort_model, pt_model, atol=1e-5)
 
+# This model with dynamic ignore_index requires torch version from commit 645119eaefd0dcf1afc54e9ad58678b5245dea78.
```
Comment (mrry) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:630]: @iK1D Does this mean that the fix also depends on a new version of PyTorch? If so, is there any way to backport it to work with PyTorch 1.8.x?

Comment (iK1D) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:630]: I am not quite farmiliar with exporter. From this commit: https://github.com/pytorch/pytorch/commit/645119eaefd0dcf1afc54e9ad58678b5245dea78#diff-a915656e4a7a07553917af566ed0f559610872d9025063d82dacb9f5c37974c9, it said to lowering NLLLoss/CrossEntropyLoss to ATen code. I think without this, we cannot override the exporter to export the CrossEntropyLoss to ATenOp, then maybe we need exporter team to change their C++ code to support that?

Comment (mrry) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:630]: If I'm reading the diff correctly, it looks like previous behavior was for `cross_entropy_loss()` to be a Python wrapper around some code that dispatches to `torch._C._nn.nll_loss()` or `torch._C._nn.nll_loss2d()`. After that commit, it dispatches directly to `torch._C._nn.cross_entropy_loss()`, which gave us the new ability to register a symbolic for it.

I *think* the way to make this work with the old version of PyTorch would be to register a similar `NegativeLogLikelihoodLossInternal` op that supports dynamic `ignore_index`, and this would give us the ability to work with PyTorch 1.8.x (and support models using NLLLoss in the future). Does that make sense?

Comment (iK1D) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:630]: Thanks @mrry It makes sense. The tricky part is, there are C++ code in exporter to fuse LogSoftmax->NegativeLogLikelihoodLoss to single SoftmaxCrossEntropyLoss. When we use NegativeLogLikelihoodLossInternal, this fuse will not work, so we also need a new transformer in our ORT side to do this fusion to reuse the kernel implementation of SoftmaxCrossEntropyLossInternal.

Comment (mrry) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:630]: Ah, that's a pity! Thanks for the explanation though.

Since it looks like a 1.9 release candidate branch is being prepared for PyTorch, we can probably just wait for the new version to be released, rather than adding a temporary new transformer in ORT.

**Review by mrry — APPROVED** [review]

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 1 above)
Comment (iK1D) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py]: I just didn't see your comment before I pushed this new commit to support torch 1.8.1... I think it's good that we can support this torch version, what if there are users working on this old version...

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 2]
```diff
@@ -2605,6 +2605,108 @@ Return true if all elements are true and false otherwise.
...
+      .Input(3, "ignore_index",
+             "Scalar tensor to specify a target value that is ignored and does not contribute to the input gradient.",
+             "I", OpSchema::Optional)
+      .Output(0, "loss", "The negative log likelihood loss", "T")
+      .TypeConstraint("T", {"tensor(float16)", "tensor(float)", "tensor(double)", "tensor(bfloat16)"},
+                      "Constrain input and output types to float tensors.")
+      .TypeConstraint("Tind", {"tensor(int32)", "tensor(int64)"}, "Constrain target to integer types")
+      .TypeConstraint("I", {"tensor(int64)"}, "Constrain ignore_index tensor to int64")
+      .TypeAndShapeInferenceFunction([](InferenceContext& ctx) { propagateElemTypeFromInputToOutput(ctx, 0, 0); })
+      .SetDoc(R"DOC(NegativeLogLikelihoodLossInternal)DOC");
```
Comment (mrry) [inline orttraining/orttraining/core/graph/training_op_defs.cc:2709]: If I understand correctly, if a user tries to use `torch.nn.NLLLoss` with the current PR, it will fail because there's no kernel for `NegativeLogLikelihoodLossInternal` - is that right?

If so, should we add a function body implementation here, based on this code for the base ONNX op?

https://github.com/onnx/onnx/blob/d08b3e951be607b9638ab84c340fabec1fdbec83/onnx/defs/math/defs.cc#L3160-L3161

...but e.g. with the `ignore_index` input passed in directly, instead of converted to a constant?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 1 above)
Comment (mrry) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py]: Oh wow, thanks for doing that 0 I didn't expect you to go to the trouble, but I appreciate it! Agreed that supporting 1.8.1 is worthwhile :).

**Comment by thiagocrepaldi** [comment]
Is this PR still needed if https://github.com/microsoft/onnxruntime/pull/7937 is merged?

**Comment by mrry** [comment]
> Is this PR still needed if #7937 is merged?

We should merge both. Even if #7937 fixes the specific issue that led to the creation of this PR, this PR strictly increases our coverage of PyTorch semantics, and may save work in the future.

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 2 above)
Comment (iK1D) [reply orttraining/orttraining/core/graph/training_op_defs.cc:2954]: Thanks @mrry for this. I've added a function body for the new Op in a new commit. But OpSet version seems a problem here. When I added a function body for an Op from com.microsoft domain and version 1, I don't know which OpSet version to use for ONNX domain, especially the Squeeze/Unsqueeze used in the function body has different input definition between OpSet12 and OpSet13. Since currently the register_custom_op is enabled for ORTModule case only, and currently the ORTModule default OpSet is version 12, I am using OpSet12 for both the function body and the UTs.

**Review by mrry — APPROVED** [review]

**Review by pengwa — DISMISSED** [review]

**Review by thiagocrepaldi — APPROVED** [review]

**Review by thiagocrepaldi — APPROVED** [review]
